File: src/VCFImpFamilyRecord.cpp
#include <map>
#include <set>
#include <cassert>
#include "../include/VCFImpFamilyRecord.h"
#include "../include/common.h"

using namespace std;

void VCFImpFamilyRecord::set_00x11_parents(size_t i, int gt) {
	const size_t	j = i == 0 ? 1 : 0;
	geno[i] = gt;		// 2 * 3 + 4
	geno[j] = 7 - gt;	// 2 * 3
	const int	prog_gt = (gt == 0) ^ (i == 0) ? 5 : 6;
	for(size_t k = 2; k < geno.size(); ++k) {
		geno[k] = prog_gt;
	}
}

int VCFImpFamilyRecord::get_records_type(const vector<RecordWithPos>& records) {
	int	type = -1;
	for(auto q = records.begin(); q != records.end(); ++q) {
		auto	*record = q->first;
		if(record->is_fixed()) {
			if(type == -1)
				type = 0;
			else if(type == 1)
				return 2;
		}
		else {
			if(type == -1)
				type = 1;
			else if(type == 0)
				return 2;
		}
	}
	return type;
}

size_t VCFImpFamilyRecord::find_fixed_index(
				const vector<std::pair<int, vector<RecordWithPos>>>& items) {
	size_t	all_fixed_index = string::npos;
	for(auto p = items.begin(); p != items.end(); ++p) {
		// 0: all fixed 1: all not fixed 2: otherwise
		const int	records_type = get_records_type(p->second);
		if(records_type == 2)
			return string::npos;
		else if(records_type == 0) {
			if(all_fixed_index == string::npos)
				all_fixed_index = p - items.begin();
			else
				return string::npos;
		}
	}
	return all_fixed_index;
}

std::pair<int, vector<VCFImpFamilyRecord::RecordWithPos>>
VCFImpFamilyRecord::which_is_fixed(const vector<RecordWithPos>& v) {
	vector<int>	gts;
	for(auto p = v.begin(); p != v.end(); ++p) {
		gts.push_back(p->first->unphased(p->second));
	}
	if(Common::is_all_same(gts))
		return make_pair(gts.front(), vector<RecordWithPos>());
	
	// collect reads in Genotype
	// ignore ./.
	map<int, vector<RecordWithPos>>	dic;
	for(size_t k = 0U; k < v.size(); ++k) {
		dic[gts[k]].push_back(v[k]);
	}
	const vector<std::pair<int, vector<RecordWithPos>>>
	items(dic.begin(), dic.end());
	
	// Is the number of the fixed records one?
	vector<bool>	bs(items.size(), true);
	for(size_t i = 0; i < items.size(); ++i) {
		const auto&	w = items[i].second;
		for(auto p = w.begin(); p != w.end(); ++p) {
			if(!p->first->is_fixed())
				bs[i] = false;
		}
	}
	int	counter = 0;
	for(auto p = bs.begin(); p != bs.end(); ++p) {
		if(*p)
			counter += 1;
	}
	if(counter != 1) {
		return make_pair(gts.front(), vector<RecordWithPos>());
	}
	
	// What number is all fixed?
	const size_t	fixed_index = find_fixed_index(items);
	if(fixed_index == string::npos)
		return make_pair(gts.front(), vector<RecordWithPos>());
	
	const int	fixed_GT = items[fixed_index].first;
	if(fixed_GT == 1)	// 0/1
		return make_pair(gts.front(), vector<RecordWithPos>());
	
	vector<RecordWithPos>	wrongs;
	for(size_t i = 0U; i < items.size(); ++i) {
		if(i == (size_t)fixed_index)
			continue;
		const auto&	records = items[i].second;
		for(auto p = records.begin(); p != records.end(); ++p) {
			if(p->first->is_00x11())
				wrongs.push_back(*p);
		}
	}
	return make_pair(fixed_GT, wrongs);
}

void VCFImpFamilyRecord::modify_00x11(const vector<RecordWithPos>& v) {
	const auto	pair = VCFImpFamilyRecord::which_is_fixed(v);
	const int	right_gt = pair.first;
	const vector<VCFImpFamilyRecord::RecordWithPos>& wrongs = pair.second;
	for(auto p = wrongs.begin(); p != wrongs.end(); ++p) {
		auto	*record = p->first;
		const size_t	i = p->second;
		record->set_00x11_parents(i, right_gt);
	}
}

File: python/VCFImpFamilyRecord.py
from __future__ import annotations

# coding: utf-8
# VCFImpFamily.py
# VCFFamilyを継承するこのプロジェクト用のクラス

from abc import ABCMeta, abstractmethod
from functools import reduce
from collections import defaultdict
from operator import add, xor
from enum import Enum

from VCFFamily import *
from TypeDeterminer import ParentComb
from common import is_all_same


#################### FillType ####################

class FillType(Enum):
	MAT = 0
	PAT = 1
	FILLED = 2
	IMPUTABLE = 3
	UNABLE = 4


#################### VCFImpFamilyRecord ####################

class VCFImpFamilyRecord(VCFFamilyRecord, metaclass=ABCMeta):
	def __init__(self, pos: int, geno: list[int], index: int,
						parents_wrong_type: str, pair: ParentComb) -> None:
		super().__init__(pos, geno)
		self.index: int = index
		self.parents_wrong_type: str = parents_wrong_type
		self.pair: ParentComb = pair
	
	def is_fixed(self) -> bool:
		return (self.pair in (ParentComb.P00x00, ParentComb.P11x11) or
				(self.pair.is_heterohomo() and self.is_right()))
	
	def is_right(self) -> bool:
		return self.parents_wrong_type == 'Right'
	
	def enable_modification(self) -> None:
		self.parents_wrong_type = 'Modifiable'
	
	# gt: 0 => 0|0 x 1|1 gt: 2 => 1|1 x 0|0
	def set_00x11_parents(self, i: int, gt: int) -> None:
		j = 1 if i == 0 else 0
		self.geno[i] = gt // 2 * 3 + 4
		self.geno[j] = 7 - gt // 2 * 3
		prog_gt = 5 if xor(gt == 0, i == 0) else 6
		for k in range(2, len(self.geno)):
			self.geno[k] = prog_gt
	
	@abstractmethod
	def is_imputable(self) -> bool:
		pass
	
	@abstractmethod
	def get_fill_type(self) -> FillType:
		pass
	
	# 複数の家系で同じサンプルを集めたRecord
	# あるGenotypeはfixedのみ、その他はfixed以外
	# -> (right GT, [(wrong record, parent index)])
	@staticmethod
	def which_is_fixed(v: list[tuple[VCFImpFamilyRecord, int]]
						) -> tuple[int, list[tuple[VCFImpFamilyRecord, int]]]:
		gts = [ r.unphased(i) for r, i in v ]
		if is_all_same(gts):
			return (gts[0], [])
		
		# collect reads in Genotype
		# ここで./.を無視してもよいはず
		dic = defaultdict(list)	# { int gt: [(record, parent index)] }
		for (r, i), gt in zip(v, gts):
			dic[gt].append((r, i))
		items = list(dic.items())	# [(int gt, [(record, parent index)])]
		
		bs = [ all(r.is_fixed() for r, i in w) for gt, w in items ]
		# 全部fixedが一つでないとあきらめる
		# 先頭のGenotypeを返すが、空のレコードのリストを返すので何でもよい
		if sum(1 for b in bs if b) != 1:
			return (gts[0], [])
		
		bs2 = [ all(not r.is_fixed() for r, i in w) for gt, w in items ]
		if sum(1 for b in bs2 if not b) != 1:
			return (gts[0], [])
		
		fixed_index = next(i for i, b in enumerate(bs) if b)
		fixed_GT = items[fixed_index][0]
		if fixed_GT == 1:	# 0/1
			return (gts[0], [])
		
		# 0/0 x 1/1パターン以外のRecordはあとで直す機会がある
		wrongs = [ (r, k) for j, item in enumerate(items) if j != fixed_index
							for r, k in item[1] if r.pair == ParentComb.P00x11 ]
		return (fixed_GT, wrongs)
	
	# 子どものGenotypeから0/0 x 1/1と推定される場合、
	# 親のどちらが0/0でどちらが1/1なのかを間違えている場合が多い
	# 親が複数の家系にまたがっているとき、修正できるかもしれない
	@staticmethod
	def modify_00x11(records: list[tuple[VCFImpFamilyRecord, int]]) -> None:
		right_gt, wrongs = VCFImpFamilyRecord.which_is_fixed(records)
		for record, i in wrongs:
			record.set_00x11_parents(i, right_gt)
